libs/utils/src/json_stream.ts

```diff
@@ -60,12 +60,24 @@ export async function* jsonStream<T>(
   input: JsonStreamInput<T>,
   { compact = true }: JsonStreamOptions = {}
 ): AsyncGenerator<string> {
-  const visitedObjects = new WeakSet();
-
   async function* jsonStreamInternal(
     value: unknown,
-    indentLevel = 0
+    indentLevel: number,
+    ancestorObjects: ReadonlySet<unknown>
   ): AsyncGenerator<string> {
+    if (ancestorObjects.has(value)) {
+      throw new Error('circular reference');
+    }
+
+    if (typeof (value as { toJSON?: () => unknown })?.toJSON === 'function') {
+      const nextAncestorObjects = new Set(ancestorObjects).add(value);
+      return yield* jsonStreamInternal(
+        (value as { toJSON(): unknown }).toJSON(),
+        indentLevel,
```

Comment:
Does it matter that you're not incrementing the `indentLevel` here? If so, why not?

Reply:
Substituting for a different value with `toJSON` doesn't create an additional level of nesting, just replaces the value with another value. We _do_ add the original value to the set of ancestors so we can properly cycle track, but the indent doesn't change.